src/drivers/magnetometer/bosch/bmm350/BMM350.cpp

```diff
@@ -0,0 +1,683 @@
+	// Set pad drive
+	RegisterWrite(Register::PAD_CTRL, (_mag_pad_drive & 0x7));
+	// Set PMU data aggregation
+	uint8_t odr = _mag_odr_mode;
+	uint8_t avg = _mag_avg_mode;
+
+	if (odr == ODR_400HZ && avg >= AVG_2) {
+		avg = AVG_NO_AVG;
+
+	} else if (odr == ODR_200HZ && avg >= AVG_4) {
+		avg = AVG_2;
+
+	} else if (odr == ODR_100HZ && avg >= AVG_8) {
+		avg = AVG_4;
+	}
+
+	uint8_t odr_reg_data = (odr & 0xf);
+	odr_reg_data = ((odr_reg_data & ~(0x30)) | ((avg << 0x4) & 0x30));
+
+	RegisterWrite(Register::PMU_CMD_AGGR_SET, odr_reg_data);
+
+	if ((readData = RegisterRead(Register::PMU_CMD_AGGR_SET)) != odr_reg_data) {
+		PX4_DEBUG("Couldn't set PMU AGGR REG");
+		success = false;
+	}
+
+	odr_reg_data = PMU_CMD_UPDATE_OAE;
+	RegisterWrite(Register::PMU_CMD, odr_reg_data);
+
+	if ((readData = RegisterRead(Register::PMU_CMD)) != odr_reg_data) {
+		PX4_DEBUG("Couldn't set PMU CMD REG");
+		success = false;
+	}
+
+	// Enable AXIS
+	uint8_t axis_data = (1 & 0x01);
+	axis_data = ((axis_data & ~(0x02)) | ((1 << 0x1) & 0x02));
+	axis_data = ((axis_data & ~(0x04)) | ((1 << 0x2) & 0x04)); // evaluates to 0x07
+
+	// PMU_CMD_AXIS_EN
+	RegisterWrite(Register::PMU_CMD_AXIS_EN, axis_data);
+
+	if ((readData = RegisterRead(Register::PMU_CMD_AXIS_EN)) != axis_data) {
+		PX4_DEBUG("Couldnt set AXIS");
+		success = false;
+	}
+
+	RegisterWrite(Register::PMU_CMD, PMU_CMD_NM);
+
+	// microTesla -> Gauss
+	_px4_mag.set_scale(0.01f);
+
+	return success;
+}
+
+int32_t BMM350::FixSign(uint32_t inval, int8_t num_bits)
+{
+	int32_t power = 1 << (num_bits - 1); // Calculate 2^(num_bits - 1)
+	int32_t retval = static_cast<int32_t>(inval);
+
+	if (retval >= power) {
+		retval -= (power << 1); // Equivalent to retval = retval - (power * 2)
+	}
+
+	return retval;
+}
+
+int8_t BMM350::ReadOutRawData(float *out_data)
+{
+	if (out_data == NULL) {
+		return -1;
+	}
+
+	float temp = 0.0;
+	struct BMM350::raw_mag_data raw_data = {0};
+	float lsb_to_ut_degc[4];
+
+	// --- Start read_uncomp_mag_temp_data
+	uint8_t mag_data[14] = {0};
+	uint8_t raw_reg[14] = {0};
+
+	uint32_t raw_mag_x, raw_mag_y, raw_mag_z, raw_temp;
+	uint8_t cmd = static_cast<uint8_t>(Register::DATAX_XLSB);
+
+	uint8_t res = transfer(&cmd, 1, (uint8_t *)&raw_reg, sizeof(raw_reg));
+
+	if (res != PX4_OK) {
+		return -1;
+	}
+
+	// Throwaway first two bytes
+	memcpy(mag_data, raw_reg + 2, sizeof(mag_data) - 2);
+
+	raw_mag_x = mag_data[0] + ((uint32_t)mag_data[1] << 8) + ((uint32_t)mag_data[2] << 16);
+	raw_mag_y = mag_data[3] + ((uint32_t)mag_data[4] << 8) + ((uint32_t)mag_data[5] << 16);
+	raw_mag_z = mag_data[6] + ((uint32_t)mag_data[7] << 8) + ((uint32_t)mag_data[8] << 16);
+	raw_temp = mag_data[9] + ((uint32_t)mag_data[10] << 8) + ((uint32_t)mag_data[11] << 16);
+
+	raw_data.raw_x = FixSign(raw_mag_x, 24);
+	raw_data.raw_y = FixSign(raw_mag_y, 24);
+	raw_data.raw_z = FixSign(raw_mag_z, 24);
+	raw_data.raw_t = FixSign(raw_temp, 24);
+	// --- End read_uncomp_mag_temp_data
+
+	// --- Start update_dafault_coefiecients
```

Comment:
This probably gets optimized away by the compiler. However, the lsb_to_ut_degc is just a vector of constant values, so it would be cleaner to simply set it as a constant vector.